openjdk/jdk
# 8283694: Improve bit manipulation and boolean to integer conversion operations on x86_64

Hi, this patch improves some operations on x86_64:

- Base variable scalar shifts have bad performance implications and should be replaced by their bmi2 counterparts if possible:
  + Bounded operands
  + Multiple uops both in fused and unfused domains
  + May result in flag stall since the operations have unpredictable flag output

- Flag to general-purpose registers operation currently uses `cmovcc`, which requires set up and 1 more spare register for constant, this could be replaced by set, which transforms the sequence:

        xorl dst, dst
        sometest
        movl tmp, 0x01
        cmovlcc dst, tmp

        into:

        xorl dst, dst
        sometest
        setbcc dst

This sequence does not need a spare register and without any drawbacks.
(Note: `movzx` does not work since move elision only occurs with different registers for input and output)

- Some small improvements:
  + Add memory variances to `tzcnt` and `lzcnt`
  + Add memory variances to `rolx` and `rorx`
  + Add missing `rolx` rules (note that `rolx dst, imm` is actually `rorx dst, size - imm`)

The speedup can be observed for variable shift instructions

        Before:
        Benchmark               (size)  Mode  Cnt   Score   Error  Units
        Integers.shiftLeft         500  avgt    5   0.836 ± 0.030  us/op
        Integers.shiftRight        500  avgt    5   0.843 ± 0.056  us/op
        Integers.shiftURight       500  avgt    5   0.830 ± 0.057  us/op
        Longs.shiftLeft            500  avgt    5   0.827 ± 0.026  us/op
        Longs.shiftRight           500  avgt    5   0.828 ± 0.018  us/op
        Longs.shiftURight          500  avgt    5   0.829 ± 0.038  us/op

        After:
        Benchmark               (size)  Mode  Cnt   Score   Error  Units
        Integers.shiftLeft         500  avgt    5   0.761 ± 0.016  us/op
        Integers.shiftRight        500  avgt    5   0.762 ± 0.071  us/op
        Integers.shiftURight       500  avgt    5   0.765 ± 0.056  us/op
        Longs.shiftLeft            500  avgt    5   0.755 ± 0.026  us/op
        Longs.shiftRight           500  avgt    5   0.753 ± 0.017  us/op
        Longs.shiftURight          500  avgt    5   0.759 ± 0.031  us/op

For `cmovcc 1, 0`, I have not been able to create a reliable microbenchmark since the benefits are mostly regarding register allocation.

Thank you very much.


---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.java.net/bylaws#reviewer))

### Issue
 * [JDK-8283694](https://bugs.openjdk.java.net/browse/JDK-8283694): Improve bit manipulation and boolean to integer conversion operations on x86_64


### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [228427b8](https://git.openjdk.java.net/jdk/pull/7968/files/228427b8d1477292e5e37b6a5c179c9111112256)
 * [Dean Long](https://openjdk.java.net/census#dlong) (@dean-long - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7968/head:pull/7968` \
`$ git checkout pull/7968`

Update a local copy of the PR: \
`$ git checkout pull/7968` \
`$ git pull https://git.openjdk.java.net/jdk pull/7968/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7968`

View PR using the GUI difftool: \
`$ git pr show -t 7968`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7968.diff">https://git.openjdk.java.net/jdk/pull/7968.diff</a>

</details>
